File: FileSave.m
//
//  FileSave.m
//
//  Created by Anthony Levings on 10/03/2014.
//  Copyright (c) 2014 Gylphi. All rights reserved. You are free to use this code but we'd grateful if you acknowledge the source when using the code or reproducing it.
//

#import "FileSave.h"

@implementation FileSave

#pragma mark - public methods

#pragma mark - data methods

+(BOOL)saveDataToDocumentsDirectory:(NSData *)fileData withName:(NSString *)path andSubDirectory:(NSString *)subdirectory
{
    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    subdirectory = [self stripSlashIfNeeded:subdirectory];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationDocumentsDirectory].path];

    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    
    // Add requested save path
    [savePath appendString:path];
    // Save the file and see if it was successful
    BOOL ok = [[NSFileManager defaultManager] createFileAtPath:[savePath copy] contents:fileData attributes:nil];
    
    // Return status of file save
    return ok;
        
}

+(BOOL)saveDataToLibraryDirectory:(NSData *)fileData withName:(NSString *)path andSubDirectory:(NSString *)subdirectory
{
    
    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    subdirectory = [self stripSlashIfNeeded:subdirectory];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationLibraryDirectory].path];
    
    
    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    
    
    // Add requested save path
    [savePath appendString:path];
    
    NSLog(@"%@",savePath);
    // Save the file and see if it was successful
    BOOL ok = [[NSFileManager defaultManager] createFileAtPath:[savePath copy] contents:fileData attributes:nil];
    
    // Return status of file save
    return ok;
}

+(BOOL)saveDataToTemporaryDirectory:(NSData *)fileData withName:(NSString *)path andSubDirectory:(NSString *)subdirectory
{
    
    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    subdirectory = [self stripSlashIfNeeded:subdirectory];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationTemporaryDirectory].path];
    
    
    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    
    
    // Add requested save path
    [savePath appendString:path];
    
    NSLog(@"%@",savePath);
    // Save the file and see if it was successful
    BOOL ok = [[NSFileManager defaultManager] createFileAtPath:[savePath copy] contents:fileData attributes:nil];
    
    // Return status of file save
    return ok;
}

+(BOOL)saveDataToCachesDirectory:(NSData *)fileData withName:(NSString *)path andSubDirectory:(NSString *)subdirectory
{
    
    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    subdirectory = [self stripSlashIfNeeded:subdirectory];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationCachesDirectory].path];
    
    
    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    
    
    // Add requested save path
    [savePath appendString:path];
    
    NSLog(@"%@",savePath);
    // Save the file and see if it was successful
    BOOL ok = [[NSFileManager defaultManager] createFileAtPath:[savePath copy] contents:fileData attributes:nil];
    
    // Return status of file save
    return ok;
}

+(BOOL)saveDataToApplicationSupportDirectory:(NSData *)fileData withName:(NSString *)path andSubDirectory:(NSString *)subdirectory excludeFromBackup:(BOOL)exclude
{
    
    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    subdirectory = [self stripSlashIfNeeded:subdirectory];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationSupportDirectory].path];
    
    
    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    
    
    // Add requested save path
    [savePath appendString:path];
    
    NSLog(@"%@",savePath);
    
    NSError *error;
    // Save the file and see if it was successful
    BOOL ok = [[NSFileManager defaultManager] createFileAtPath:[savePath copy] contents:fileData attributes:nil];
    if (exclude) { [[NSURL URLWithString:savePath] setResourceValue:@YES
                                                               forKey:NSURLIsExcludedFromBackupKey
                                                                error:&error];}
    if (error) NSLog(@"%@", error);
    
    // Return status of file save
    return ok;
}

#pragma mark - string methods

+(BOOL)saveContentsOfStringToDocumentsDirectory:(NSString *)fileString withName:(NSString *)path andSubDirectory:(NSString *)subdirectory {
    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationDocumentsDirectory].path];
    
    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    // Add requested save path
    [savePath appendString:path];
    
    NSError *error;
    // Save the file and see if it was successful
    BOOL ok = [fileString writeToFile:savePath atomically:NO encoding:NSUTF8StringEncoding error:&error];
    
    if (error) NSLog(@"%@", error);
    
    // Return status of file save
    return ok;

}
+(BOOL)saveContentsOfStringToLibraryDirectory:(NSString *)fileString withName:(NSString *)path andSubDirectory:(NSString *)subdirectory {

    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationLibraryDirectory].path];
    
    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    
    // Add requested save path
    [savePath appendString:path];
    
    NSError *error;
    // Save the file and see if it was successful
    BOOL ok = [fileString writeToFile:savePath atomically:NO encoding:NSUTF8StringEncoding error:&error];
    
    if (error) NSLog(@"%@", error);
    
    // Return status of file save
    return ok;

}
+(BOOL)saveContentsOfStringToTemporaryDirectory:(NSString *)fileString withName:(NSString *)path andSubDirectory:(NSString *)subdirectory {
    
    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationTemporaryDirectory].path];
    
    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    
    // Add requested save path
    [savePath appendString:path];
    
    NSError *error;
    // Save the file and see if it was successful
    BOOL ok = [fileString writeToFile:savePath atomically:NO encoding:NSUTF8StringEncoding error:&error];
    
    if (error) NSLog(@"%@", error);
    
    // Return status of file save
    return ok;
    
}

+(BOOL)saveContentsOfStringToCachesDirectory:(NSString *)fileString withName:(NSString *)path andSubDirectory:(NSString *)subdirectory {
    
    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationCachesDirectory].path];
    
    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    
    // Add requested save path
    [savePath appendString:path];
    
    NSError *error;
    // Save the file and see if it was successful
    BOOL ok = [fileString writeToFile:savePath atomically:NO encoding:NSUTF8StringEncoding error:&error];
    
    if (error) NSLog(@"%@", error);
    
    // Return status of file save
    return ok;
    
}


+(BOOL)saveContentsOfStringToApplicationSupportDirectory:(NSString *)fileString withName:(NSString *)path andSubDirectory:(NSString *)subdirectory excludeFromBackUp:(BOOL)exclude {
    
    // Remove unnecessary slash if need
    path = [self stripSlashIfNeeded:path];
    
    // Create generic beginning to file save path
    NSMutableString *savePath = [[NSMutableString alloc] initWithFormat:@"%@/",[self applicationSupportDirectory].path];
    
    if (subdirectory){
        [savePath appendString:subdirectory];
        [self createSubDirectory:[savePath copy]];
        [savePath appendString:@"/"];
    }
    
    // Add requested save path
    [savePath appendString:path];
    
    NSError *error;
    // Save the file and see if it was successful
    BOOL ok = [fileString writeToFile:savePath atomically:NO encoding:NSUTF8StringEncoding error:&error];
    
    if (error) NSLog(@"%@", error);
    
    
    if (exclude) { [[NSURL URLWithString:savePath] setResourceValue:@YES
                                        forKey:NSURLIsExcludedFromBackupKey
                                                                error:&error];}
    if (error) NSLog(@"%@", error);
    // Return status of file save
    return ok;
    
}
#pragma mark - private methods

#pragma mark - directories

+ (NSURL *)applicationDocumentsDirectory {
    
    NSString *documentsDirectory;
NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
if ([paths count] > 0) {
    documentsDirectory = [paths objectAtIndex:0];
}

    return [NSURL URLWithString:documentsDirectory];
}

+ (NSURL *)applicationLibraryDirectory {
 NSString *libraryDirectory;
NSArray *paths = NSSearchPathForDirectoriesInDomains(NSLibraryDirectory, NSUserDomainMask, YES);
if ([paths count] > 0) {
    libraryDirectory = [paths objectAtIndex:0];
}   
    
    return [NSURL URLWithString:libraryDirectory];
}

+(NSURL *)applicationSupportDirectory {
 NSString *applicationSupportDirectory;
NSArray *paths = NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
if ([paths count] > 0) {
    applicationSupportDirectory = [paths objectAtIndex:0];
}   
    
    return [NSURL URLWithString:applicationSupportDirectory];
}   
    
    



+(NSURL *)applicationTemporaryDirectory {
    
    NSString *temporaryDirectory = NSTemporaryDirectory();
    
    return [NSURL URLWithString:temporaryDirectory];
    
}
+(NSURL *)applicationCachesDirectory {
    NSString *cachesDirectory;
NSArray *paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
if ([paths count] > 0) {
    cachesDirectory = [paths objectAtIndex:0];
}   
    
    return [NSURL URLWithString:cachesDirectory];
}
#pragma mark - strip slashes

+(NSString *)stripSlashIfNeeded:(NSString *)stringWithPossibleSlash {
    
    // If the file name contains a slash at the beginning then we remove so that we don't end up with two
    if ([stringWithPossibleSlash compare:@"/" options:NSLiteralSearch range:NSMakeRange(0, 1)]==NSOrderedSame) {
        stringWithPossibleSlash = [stringWithPossibleSlash substringFromIndex:1];
    }
    // Return the string with no slash at the beginning
    return stringWithPossibleSlash;
}

+(BOOL)createSubDirectory:(NSString *)subdirectoryPath {
    NSError *error;
    BOOL isDir;
    BOOL exists = [[NSFileManager defaultManager] fileExistsAtPath:subdirectoryPath isDirectory:&isDir];
    if (exists) {
        /* a file of the same name exists, we don't care about this so won't do anything */
        if (isDir) {
            /* subdirectory already exists, don't create it again */
            return YES;
        }
    }
    BOOL success = [[NSFileManager defaultManager] createDirectoryAtPath:subdirectoryPath withIntermediateDirectories:YES attributes:nil error:&error];
    
    if (error) NSLog(@"%@",error);
    
    return success;
}


@end

File: FileSave.swift
import Foundation

class FileSave {
    
    class func saveDataToDocumentsDirectory(fileData:NSData, path:String, subdirectory:String?) -> Bool
    {
    // Remove unnecessary slash if need
    var newPath = FileSaver.stripSlashIfNeeded(path)
        var newSubdirectory:String?
        if subdirectory {
        newSubdirectory = FileSaver.stripSlashIfNeeded(subdirectory!)
        }
    // Create generic beginning to file save path
    var savePath = self.applicationDocumentsDirectory().path+"/"
    
    if newSubdirectory {
    savePath += newSubdirectory!
    self.createSubDirectory(savePath)
    savePath += "/"
    }
    
    // Add requested save path
    savePath += newPath
        
        println(savePath)
    // Save the file and see if it was successful
        var ok:Bool = NSFileManager.defaultManager().createFileAtPath(savePath,contents:fileData, attributes:nil)
    
    // Return status of file save
    return ok;
    
    }
    
    class func saveDataToLibraryDirectory(fileData:NSData, path:String, subdirectory:String?) -> Bool
    {
    
        // Remove unnecessary slash if need
        var newPath = self.stripSlashIfNeeded(path)
        var newSubdirectory:String?
        if subdirectory {
            newSubdirectory = FileSaver.stripSlashIfNeeded(subdirectory!)
        }
        
        // Create generic beginning to file save path
        var savePath = self.applicationLibraryDirectory().path+"/"
        
        if newSubdirectory {
            savePath += subdirectory!
            self.createSubDirectory(savePath)
            savePath += "/"
        }
        
        // Add requested save path
        savePath += newPath
        // Save the file and see if it was successful
        var ok:Bool = NSFileManager.defaultManager().createFileAtPath(savePath,contents:fileData, attributes:nil)
        
        // Return status of file save
        return ok;
    }
    
    class func saveDataToTemporaryDirectory(fileData:NSData, path:String, subdirectory:String?) -> Bool
    {
    
    // Remove unnecessary slash if need
    var newPath = self.stripSlashIfNeeded(path)
        var newSubdirectory:String?
        if subdirectory {
            newSubdirectory = FileSaver.stripSlashIfNeeded(subdirectory!)
        }
    
    // Create generic beginning to file save path
    var savePath = self.applicationTemporaryDirectory().path+"/"
    
    if newSubdirectory {
    savePath += newSubdirectory!
    self.createSubDirectory(savePath)
    savePath += "/"
    }
    
    // Add requested save path
    savePath += path
    // Save the file and see if it was successful
    var ok:Bool = NSFileManager.defaultManager().createFileAtPath(savePath,contents:fileData, attributes:nil)
    
    // Return status of file save
    return ok;
    }
    
    class func saveDataToCachesDirectory(fileData:NSData, path:String, subdirectory:String?) -> Bool
    {
    
        
        // Remove unnecessary slash if need
        var newPath = self.stripSlashIfNeeded(path)
        var newSubdirectory:String?
        if subdirectory {
            newSubdirectory = FileSaver.stripSlashIfNeeded(subdirectory!)
        }
        
        // Create generic beginning to file save path
        var savePath = self.applicationCachesDirectory().path+"/"
        
        if newSubdirectory {
            savePath += newSubdirectory!
            self.createSubDirectory(savePath)
            savePath += "/"
        }
        
        // Add requested save path
        savePath += path
        // Save the file and see if it was successful
        var ok:Bool = NSFileManager.defaultManager().createFileAtPath(savePath,contents:fileData, attributes:nil)
        
        // Return status of file save
        return ok;
    }
    
    class func saveDataToApplicationSupportDirectory(fileData:NSData, path:String, subdirectory:String?, excludeFromBackup:Bool) -> Bool
    {
    
    // Remove unnecessary slash if need
    var newPath = self.stripSlashIfNeeded(path)
        var newSubdirectory:String?
        if subdirectory {
            newSubdirectory = FileSaver.stripSlashIfNeeded(subdirectory!)
        }
    
    // Create generic beginning to file save path
    var savePath = self.applicationSupportDirectory().path+"/"
    
    if newSubdirectory {
    savePath += newSubdirectory!
    self.createSubDirectory(savePath)
    savePath += "/"
    }
    
    // Add requested save path
    savePath += path
    // Save the file and see if it was successful
    var ok:Bool = NSFileManager.defaultManager().createFileAtPath(savePath,contents:fileData, attributes:nil)
    
    var error:NSError?
        
    if excludeFromBackup {
        NSURL.URLWithString(savePath).setResourceValue(true,forKey:NSURLIsExcludedFromBackupKey,
    error:&error)
        }
        if error {
         println(error);
        }
    
    // Return status of file save
    return ok;
    }
    
// string methods
    
    class func saveContentsOfStringToDocumentsDirectory(fileString:String, path:String, subdirectory:String) -> Bool {
    // Remove unnecessary slash if need
    var newPath = self.stripSlashIfNeeded(path)
        var newSubdirectory:String? = self.stripSlashIfNeeded(subdirectory)
    
    // Create generic beginning to file save path
    var savePath = self.applicationDocumentsDirectory().path+"/"
    
    if newSubdirectory {
    savePath += newSubdirectory!
    self.createSubDirectory(savePath)
    savePath += "/"
    }
    
    // Add requested save path
    savePath += path
    
    var error:NSError?;
    // Save the file and see if it was successful
        var ok:Bool = fileString.writeToFile(savePath, atomically:false, encoding:NSUTF8StringEncoding, error:&error)
    
        if error {println(error)}
    
    // Return status of file save
    return ok;
    
    }
    class func saveContentsOfStringToLibraryDirectory(fileString:String, path:String, subdirectory:String) -> Bool {
    
        var newPath = self.stripSlashIfNeeded(path)
        var newSubdirectory:String? = self.stripSlashIfNeeded(subdirectory)
        
        // Create generic beginning to file save path
        var savePath = self.applicationLibraryDirectory().path+"/"
        
        if newSubdirectory {
            savePath += subdirectory
            self.createSubDirectory(savePath)
            savePath += "/"
        }
        
        // Add requested save path
        savePath += path
        
        var error:NSError?;
        // Save the file and see if it was successful
        var ok:Bool = fileString.writeToFile(savePath, atomically:false, encoding:NSUTF8StringEncoding, error:&error)
        
        if (error) {println(error)}
        
        // Return status of file save
        return ok;
    
    }
    class func saveContentsOfStringToTemporaryDirectory(fileString:String, path:String, subdirectory:String) -> Bool {
    
    var newPath = self.stripSlashIfNeeded(path)
        var newSubdirectory:String? = self.stripSlashIfNeeded(subdirectory)
    
    // Create generic beginning to file save path
    var savePath = self.applicationTemporaryDirectory().path+"/"
    
    if newSubdirectory {
    savePath += newSubdirectory!
    self.createSubDirectory(savePath)
    savePath += "/"
    }
    
    // Add requested save path
    savePath += path
    
    var error:NSError?;
    // Save the file and see if it was successful
        var ok:Bool = fileString.writeToFile(savePath, atomically:false, encoding:NSUTF8StringEncoding, error:&error)
    
    if error {
        println(error)
        }
    
    // Return status of file save
    return ok;

    }
    
    class func saveContentsOfStringToCachesDirectory(fileString:String, path:String, subdirectory:String) -> Bool {
    
    var newPath = self.stripSlashIfNeeded(path)
        var newSubdirectory:String? = self.stripSlashIfNeeded(subdirectory)
    
    // Create generic beginning to file save path
    var savePath = self.applicationCachesDirectory().path+"/"
    
    if newSubdirectory {
    savePath += newSubdirectory!
    self.createSubDirectory(savePath)
    savePath += "/"
    }
    
    // Add requested save path
    savePath += path
    
    var error:NSError?;
    // Save the file and see if it was successful
        var ok:Bool = fileString.writeToFile(savePath, atomically:false, encoding:NSUTF8StringEncoding, error:&error)
    
    if error {
        println(error)
        }
    
    // Return status of file save
    return ok;

    
    }
    
    
    class func saveContentsOfStringToApplicationSupportDirectory(fileString:String, path:String, subdirectory:String, excludeFromBackUp:Bool) -> Bool {
    
    var newPath = self.stripSlashIfNeeded(path)
        var newSubdirectory:String? = self.stripSlashIfNeeded(subdirectory)
    
    // Create generic beginning to file save path
    var savePath = self.applicationSupportDirectory().path+"/"
    
    if newSubdirectory {
    savePath += newSubdirectory!
    self.createSubDirectory(savePath)
    savePath += "/"
    }
    
    // Add requested save path
    savePath += path
    
    var error:NSError?;
    // Save the file and see if it was successful
    var ok:Bool = fileString.writeToFile(savePath, atomically:false, encoding:NSUTF8StringEncoding, error:&error)
    
    if error {println(error)}
    
    var errorExclude:NSError?
    
    if excludeFromBackUp {
        NSURL.URLWithString(savePath).setResourceValue(true, forKey:NSURLIsExcludedFromBackupKey,
    error:&error)
    }
    if errorExclude {
    println(errorExclude);
    }
    
    // Return status of file save
    return ok;
    }

    
    
// private methods
    
  //directories
    
 class func applicationDocumentsDirectory() -> NSURL {
    
        var documentsDirectory:String?
        var paths:AnyObject[] = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true);
    if paths.count > 0 {
        if let pathString = paths[0] as? NSString {
            documentsDirectory = pathString
        }
    }
    
    return NSURL.URLWithString(documentsDirectory)
    }
    
   class func applicationLibraryDirectory() -> NSURL {
        var libraryDirectory:String?
        var paths:AnyObject[] = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.LibraryDirectory, NSSearchPathDomainMask.UserDomainMask, true);
        if paths.count > 0 {
            if let pathString = paths[0] as? NSString {
                libraryDirectory = pathString
            }
        }
    return NSURL.URLWithString(libraryDirectory)
    }
    
class func applicationSupportDirectory() -> NSURL {
    var applicationSupportDirectory:String?
    var paths:AnyObject[] = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.ApplicationSupportDirectory, NSSearchPathDomainMask.UserDomainMask, true);
    if paths.count > 0 {
        if let pathString = paths[0] as? NSString {
            applicationSupportDirectory = pathString
        }
    }
    return NSURL.URLWithString(applicationSupportDirectory)
    }
    
    
    
    
    
  class func applicationTemporaryDirectory() -> NSURL {
    
        var temporaryDirectory:String? = NSTemporaryDirectory();
    
    return NSURL.URLWithString(temporaryDirectory)
    
    }
    class func applicationCachesDirectory() -> NSURL {
        
        var cachesDirectory:String?

        var paths = NSSearchPathForDirectoriesInDomains(.ApplicationSupportDirectory,.UserDomainMask, true);
        
        if paths.count > 0 {
            if let pathString = paths[0] as? NSString {
                cachesDirectory = pathString
            }
        }
    return NSURL.URLWithString(cachesDirectory);
    }
   //pragma mark - strip slashes
    
class func stripSlashIfNeeded(stringWithPossibleSlash:String) -> String {
    var stringWithoutSlash:String = stringWithPossibleSlash
    // If the file name contains a slash at the beginning then we remove so that we don't end up with two
    if stringWithPossibleSlash.hasPrefix("/") {
    stringWithoutSlash = stringWithPossibleSlash.substringFromIndex(1)
    }
    // Return the string with no slash at the beginning
    return stringWithoutSlash
    }
    
    class func createSubDirectory(subdirectoryPath:NSString) -> Bool {
    var error:NSError?
    var isDir:ObjCBool=false;
    var exists:Bool = NSFileManager.defaultManager().fileExistsAtPath(subdirectoryPath, isDirectory:&isDir)
    if (exists) {
    /* a file of the same name exists, we don't care about this so won't do anything */
    if isDir==true {
    /* subdirectory already exists, don't create it again */
    return true;
    }
    }
    var success:Bool = NSFileManager.defaultManager().createDirectoryAtPath(subdirectoryPath, withIntermediateDirectories:true, attributes:nil, error:&error)
    
    if error { println(error) }
    
    return success;
    }
}
